dev/NavigationView/NavigationView.cpp

```diff
@@ -379,10 +427,354 @@ void NavigationView::OnApplyTemplate()
+        UpdateLeftRepeaterItemSource(dataSource);
+    }
+
+    if (IsTopNavigationView())
+    {
+        InvalidateTopNavPrimaryLayout();
+        UpdateSelectedItem();
+    }
+}
+
+void NavigationView::UpdateLeftRepeaterItemSource(const winrt::IInspectable& items)
+{
+    UpdateItemsRepeaterItemsSource(m_leftNavRepeater.get(), items);
+}
+
+void NavigationView::UpdateTopNavRepeatersItemSource(const winrt::IInspectable& items)
+{
+    // Unhook TopNav repeater
+    UpdateItemsRepeaterItemsSource(m_topNavRepeater.get(), nullptr);
+    UpdateItemsRepeaterItemsSource(m_topNavRepeaterOverflowView.get(), nullptr);
+
+    // Change data source and setup vectors
+    m_topDataProvider.SetDataSource(items);
+
+    // rebinding
+    if (items)
+    {
+        UpdateItemsRepeaterItemsSource(m_topNavRepeater.get(), m_topDataProvider.GetPrimaryItems());
+        UpdateItemsRepeaterItemsSource(m_topNavRepeaterOverflowView.get(), m_topDataProvider.GetOverflowItems());
+    }
+    else
+    {
+        UpdateItemsRepeaterItemsSource(m_topNavRepeater.get(), nullptr);
+        UpdateItemsRepeaterItemsSource(m_topNavRepeaterOverflowView.get(), nullptr);
+    }
+}
+
+void NavigationView::UpdateItemsRepeaterItemsSource(const winrt::ItemsRepeater& ir,
+    const winrt::IInspectable& itemsSource)
+{
+    if (ir)
+    {
+        auto oldItemsSource = ir.ItemsSource();
+        if (oldItemsSource != itemsSource)
+        {
+            ir.ItemsSource(itemsSource);
+        }
+    }
+}
+
+void NavigationView::OnNavigationViewItemIsSelectedPropertyChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args)
+{
+    auto nvib = sender.try_as<winrt::NavigationViewItemBase>();
+    if (nvib)
+    {
+        // Check whether the container that triggered this call back is the selected container
+        bool newItemIsSelectedItem = IsContainerTheSelectedItemInTheSelectionModel(nvib);
+        bool newIsSelected = nvib.IsSelected();
+
+        if (newIsSelected && !newItemIsSelectedItem)
+        {
+            winrt::IndexPath ip = GetIndexPathForContainer(nvib);
+            m_selectionModel.SelectAt(ip);
+        }
+        else if (!newIsSelected && newItemIsSelectedItem)
+        {
+            auto indexPath = GetIndexPathForContainer(nvib);
+            auto indexPathFromModel = m_selectionModel.SelectedIndex();
+
+            if (indexPath && indexPathFromModel && indexPath.GetAt(0) == indexPathFromModel.GetAt(0))
+            {
+                m_selectionModel.DeselectAt(indexPath);
+            }
+        }
+    }
+}
+
+void NavigationView::RaiseItemInvokedForNavigationViewItem(const winrt::NavigationViewItem& nvi)
+{
+    winrt::IInspectable nextItem = nullptr;
+    auto prevItem = SelectedItem();
+    auto parentIR = GetParentItemsRepeaterForContainer(nvi);
+
+    bool isInOverflow = parentIR.Name() == c_overflowRepeater;
+    bool itemSelectsOnInvoked = nvi.SelectsOnInvoked();
+
+    auto itemIndex = parentIR.GetElementIndex(nvi);
```

Comment:
Move this inside the if block below.

Reply:
Per c++ guidelines ES.5 and ES.6 [https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es5-keep-scopes-small](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es5-keep-scopes-small)

---
In reply to: [352760263](https://github.com/microsoft/microsoft-ui-xaml/pull/1683#discussion_r352760263) [](ancestors = 352760263)